packages/common/src/router-v2.js

```diff
@@ -0,0 +1,86 @@
+const { getRoutePaths } = require('./router');
+const { Router } = require('express');
+
+/**
+ * @typedef {import('express').IRouter} IRouter
+ * @typedef {import('express').Handler} Handler
+ * @typedef {Object<string, IRouter>} RouteDict
+ * @typedef {import('./router-v2-types').HttpMethods} HttpMethods
+ * @typedef {import('./router-v2-types').RouterModule} RouterModule
+ * @typedef {{ includeRoot?: boolean, backwardsCompatibilityModeEnabled?: boolean }} Options
+ */
+
+/** @type {Array<HttpMethods>} */
+const HttpMethods = [
+	'connect',
+	'delete',
+	'get',
+	'head',
+	'options',
+	'patch',
+	'post',
+	'put',
+	'trace'
+];
+
+/** @type {(str: *) => str is HttpMethods} */
+const stringIsRecognisedExport = (str) => HttpMethods.includes(str);
+
+/**
+ * @param {string} [directory]
+ * @param {Options} [options]
+ * @returns {IRouter}
+ */
+exports.getRouter = (
+	directory = __dirname,
+	{ backwardsCompatibilityModeEnabled = false, ...options } = {
+		includeRoot: false,
+		backwardsCompatibilityModeEnabled: false
+	}
+) =>
+	getRoutePaths(directory, options)
+		.sort((a, b) => b.split('/').length - a.split('/').length)
+		.reduce((router, dirName) => {
+			/** @type {RouterModule} */
+			const { middleware, ...module } = require(`${dirName}`);
+			const relativePath = dirName
+				.replace(new RegExp(`^${directory}/?`), '/') // just need relative path
+				.replace(/_/g, ':') // need ':param' but Windows doesn't like ':' in folder names so we use '_param'
+				.replace('/index.js', '');
+
+			if (middleware?.[0]) {
+				router.use(relativePath, ...middleware[0]);
+			}
+
+			Object.entries(module).forEach(([method, handler]) => {
+				if (backwardsCompatibilityModeEnabled && method === 'router') {
+					router.use(relativePath, handler); // in this instance "handler" should actually be a router
+					return;
+				}
+				if (!stringIsRecognisedExport(method)) {
+					// console.warn(
+					// 	`Skipping ${method} function exported by ${dirName}/index.js as the function name is not a recognised HTTP method.`
```

Comment:
I don't know how easy this would be...

but would be great to have a single log of all the routes/middleware that are applied, and in what order, would be super useful with debugging

e.g. routes
/test generic-middleware['', '']
/test :get 
/test :post middleware ['require-login']
/test/_param/ :get

also could bring this warning back

pass through an optional logger to be passed into getRouter and use it if present

Reply:
Sounds good

Reply:
This did take a bit of work but I think it was worth it

Forms web app output: 
<img width="975" alt="Screenshot 2024-07-19 at 15 08 32" src="https://github.com/user-attachments/assets/3e87d03e-a2f2-49e6-8509-5ea0733fff0f">

[](url)